inference-engine/src/vpu/common/src/ngraph/transformations/merge_gather_gather_elements.cpp

```diff
@@ -0,0 +1,156 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "vpu/ngraph/transformations/merge_gather_gather_elements.hpp"
+
+#include <vpu/ngraph/operations/exp_gather_elements.hpp>
+#include <vpu/ngraph/utilities.hpp>
+
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/opsets/opset6.hpp>
+#include <ngraph/pattern/matcher.hpp>
+#include <ngraph/pattern/op/label.hpp>
+#include <ngraph/validation_util.hpp>
+#include <ngraph/log.hpp>
+
+#include <numeric>
+
+NGRAPH_RTTI_DEFINITION(vpu::MergeGatherGatherElements, "MergeGatherGatherElements", 0);
+
+namespace vpu {
+
+bool MergeGatherGatherElements::run_on_function(std::shared_ptr<ngraph::Function> f) {
+    bool wasGraphChanged = false;
+
+    const auto gatherData = ngraph::pattern::any_input();
+    const auto gatherIndices = ngraph::pattern::any_input();
+    const auto gather = ngraph::pattern::wrap_type<ngraph::opset6::Gather>(
+        {gatherData, gatherIndices, ngraph::pattern::any_input()},
+        ngraph::pattern::consumers_count(1));
+
+    const auto squeezeAxes = ngraph::pattern::any_input(
+        ngraph::pattern::op::as_value_predicate(ngraph::pattern::has_class<ngraph::opset6::Constant>()));
+    const auto squeeze = ngraph::pattern::wrap_type<ngraph::opset6::Squeeze>({gather, squeezeAxes}, ngraph::pattern::consumers_count(1));
+
+    const auto transposePerm = ngraph::pattern::any_input(
+        ngraph::pattern::op::as_value_predicate(ngraph::pattern::has_class<ngraph::opset6::Constant>()));
+    const auto transpose = ngraph::pattern::wrap_type<ngraph::opset6::Transpose>({squeeze, transposePerm});
+
+    const auto m = std::make_shared<ngraph::pattern::Matcher>(transpose, "GatherSqueezeTransposeMatcher");
+    for (const auto& node : f->get_ordered_ops()) {
+        if (!m->match(node)) {
+            continue;
+        }
+
+        auto& patternMap = m->get_pattern_value_map();
+
+        std::vector<ngraph::Node*> shapeOfs;
+        std::vector<ngraph::Node*> gatherElements;
+
+        const auto& matchedTranspose = patternMap.at(transpose);
+        const auto& transposeConsumers = matchedTranspose.get_target_inputs();
+
+        const auto isShapeOfOrGatherElements = [](const ngraph::Input<ngraph::Node>& consumer) {
+            return consumer.get_node()->get_type_info() == ngraph::opset6::ShapeOf::type_info ||
+                   consumer.get_node()->get_type_info() == ngraph::opset6::GatherElements::type_info;
+        };
```

Comment:
Use is_castable to avoid issues when matching TypeRelaxed operations. 
```suggestion
        const auto isShapeOfOrGatherElements = [](const ngraph::Input<ngraph::Node>& consumer) {
            return consumer.get_node()->get_type_info().is_castable(ngraph::opset6::ShapeOf::type_info) ||
                   consumer.get_node()->get_type_info().is_castable(ngraph::opset6::GatherElements::type_info);
        };
```

Reply:
Ok, fixed.